src/engine/ExecuteUpdate.cpp

```diff
@@ -2,57 +2,61 @@
-                                              std::move(graphUpdate.toInsert_));
-  auto [toDeleteTemplates, localVocabDelete] =
-      ExecuteUpdate::transformTriplesTemplate(vocab, qet.getVariableColumns(),
-                                              std::move(graphUpdate.toDelete_));
+  auto prepareTemplateAndResultContainer =
+      [&vocab, &qet,
+       &res](std::vector<SparqlTripleSimpleWithGraph>&& tripleTemplates) {
+        auto [transformedTripleTemplates, localVocab] =
+            transformTriplesTemplate(vocab, qet.getVariableColumns(),
+                                     std::move(tripleTemplates));
+        std::vector<IdTriple<>> updateTriples;
+        // The maximum result size is size(query result) x num template rows.
+        // The actual result can be smaller if there are template rows with
+        // variables for which a result row does not have a value.
+        updateTriples.reserve(res->idTable().size() *
+                              transformedTripleTemplates.size());
+
+        return std::tuple{std::move(transformedTripleTemplates),
+                          std::move(updateTriples), std::move(localVocab)};
+      };
+
+  auto [toInsertTemplates, toInsert, localVocabInsert] =
+      prepareTemplateAndResultContainer(std::move(graphUpdate.toInsert_));
+  auto [toDeleteTemplates, toDelete, localVocabDelete] =
+      prepareTemplateAndResultContainer(std::move(graphUpdate.toDelete_));
 
-  std::vector<IdTriple<>> toInsert;
-  std::vector<IdTriple<>> toDelete;
-  // The maximum result size is size(query result) x num template rows. The
-  // actual result can be smaller if there are template rows with variables for
-  // which a result row does not have a value.
-  toInsert.reserve(res->idTable().size() * toInsertTemplates.size());
-  toDelete.reserve(res->idTable().size() * toDeleteTemplates.size());
   for (const auto& [pair, range] :
        ExportQueryExecutionTrees::getRowIndices(query._limitOffset, *res)) {
     auto& idTable = pair.idTable_;
-    for (uint64_t i : range) {
-      ExecuteUpdate::computeAndAddQuadsForResultRow(toInsertTemplates, toInsert,
-                                                    idTable, i);
+    for (const uint64_t i : range) {
+      computeAndAddQuadsForResultRow(toInsertTemplates, toInsert, idTable, i);
       cancellationHandle->throwIfCancelled();
 
-      ExecuteUpdate::computeAndAddQuadsForResultRow(toDeleteTemplates, toDelete,
-                                                    idTable, i);
+      computeAndAddQuadsForResultRow(toDeleteTemplates, toDelete, idTable, i);
       cancellationHandle->throwIfCancelled();
     }
   }
 
-  // TODO<qup42> use the actual DeltaTriples object
-  DeltaTriples deltaTriples(index);
-  deltaTriples.insertTriples(cancellationHandle, std::move(toInsert));
   deltaTriples.deleteTriples(cancellationHandle, std::move(toDelete));
+  deltaTriples.insertTriples(cancellationHandle, std::move(toInsert));
```

Comment:
Are there queries that do insert and delete at the same time?
If so, which is the correct order? Explicit comment this here and make sure that there is a unit test.

Reply:
Queries like `DELETE { <a> <b> <c> } INSERT { <a> <b> <c> } WHERE { ?s <is-a> ?o }` do both in a single update. The correct order is the intuitive one - first delete then insert. There is a test with the aforementioned update that checks that the triples is inserted and not deleted in the end.